pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/BigNumberOfSegmentsIntegrationTest.java

```diff
@@ -0,0 +1,264 @@
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.config.table.TableType;
+import org.apache.pinot.spi.data.FieldSpec;
+import org.apache.pinot.spi.data.Schema;
+import org.apache.pinot.spi.utils.TimeUtils;
+import org.apache.pinot.spi.utils.builder.TableConfigBuilder;
+import org.apache.pinot.util.TestUtils;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import static org.apache.pinot.integration.tests.ClusterIntegrationTestUtils.getBrokerQueryApiUrl;
+
+
+/**
+ * Create many small segments with many columns to test metadata overhead.
+ */
+public class BigNumberOfSegmentsIntegrationTest extends BaseClusterIntegrationTestSet {
+
+  static final int FILES_NO = 1000;
+  static final int RECORDS_NO = 5;
+  static final String INT_COL = "i";
+  static final String LONG_COL = "j";
+  static final String STR_COL_PREFIX = "s";
+  static final String FLOAT_COL = "f";
+  static final String DOUBLE_COL = "d";
+  static final int STR_COL_NUM = 200;
+
+  @BeforeClass
+  public void setUp()
+      throws Exception {
+    System.out.println("Pid is " + ProcessHandle.current().pid());
+    TestUtils.ensureDirectoriesExistAndEmpty(_tempDir, _segmentDir, _tarDir);
+
+    startZk();
+    startKafka();
+    startController();
+    startServer();
+    startBroker();
+
+    Schema.SchemaBuilder builder = new Schema.SchemaBuilder()
+        .setSchemaName(DEFAULT_SCHEMA_NAME)
+        .addSingleValueDimension(INT_COL, FieldSpec.DataType.INT)
+        .addSingleValueDimension(LONG_COL, FieldSpec.DataType.LONG)
+        .addSingleValueDimension(FLOAT_COL, FieldSpec.DataType.FLOAT)
+        .addSingleValueDimension(DOUBLE_COL, FieldSpec.DataType.DOUBLE)
+        .addDateTimeField(DEFAULT_TIME_COLUMN_NAME, FieldSpec.DataType.TIMESTAMP, "TIMESTAMP", "1:MILLISECONDS");
+
+    for (int i = 0; i < STR_COL_NUM; i++) {
+      builder.addSingleValueDimension(STR_COL_PREFIX + i, FieldSpec.DataType.STRING);
+    }
+
+    Schema schema = builder.build();
+
+    List<File> avroFiles = createAvroFiles(_tempDir);
+
+    addSchema(schema);
+    TableConfig tableConfig = createOfflineTableConfig(schema);
+    addTableConfig(tableConfig);
+
+    ClusterIntegrationTestUtils.buildSegmentsFromAvro(avroFiles, tableConfig, schema, 0, _segmentDir, _tarDir);
+    uploadSegments(DEFAULT_TABLE_NAME, _tarDir);
+
+    // Wait for all documents loaded
+    TestUtils.waitForCondition(() -> getCurrentCountStarResult(DEFAULT_TABLE_NAME) == FILES_NO * RECORDS_NO, 100L,
+        600_000,
+        "Failed to load  documents", true, Duration.ofMillis(60_000 / 10));
+
+    setUseMultiStageQueryEngine(true);
+  }
+
+  // to slow for CI (1+ minute)
+  @Test(enabled = false)
+  public void testCreateManySegments()
+      throws Exception {
+    JsonNode node = postV2Query("select sum(i) + sum(j) + sum(d), count(*) from " + DEFAULT_TABLE_NAME);
+
+    assertNoError(node);
+  }
```

Comment:
nit: IICU no test from this class will be executed on CI, am I right? Can we add that into the class javadoc?

Reply:
Done